source/extensions/common/wasm/context.cc

```diff
@@ -1407,7 +1407,8 @@ void Context::onGrpcReceiveTrailingMetadata(uint32_t token, Http::HeaderMapPtr&&
 }
 
 WasmResult Context::defineMetric(MetricType type, absl::string_view name, uint32_t* metric_id_ptr) {
-  auto stat_name = wasm_->stat_name_set_->add(name);
+  Stats::StatNameManagedStorage storage(name, wasm()->scope_->symbolTable());
```

Comment:
What's the lifetime of wasm()->scope_? Should we create a scope for the VM instead so you can drop stats that are no longer used in the VM?

I'm pretty sure you can have multiple Scope objects with the same name, and that the underlying stats will be shared if they have equal StatNames.

Reply:
I wasn't sure how the storage worked, but we have 1 VM per worker which have very similar lifetimes and Stats.  I am creating a single Scope and giving a std::shared_ptr to each of the VMs so that when the last one is deleted the Scope is deleted.  We could allow individual VMs to purge all stats if we gave each one a separate Scope.   However, we currently do not have that API, so it isn't clear to me what the tradeoff would be between sharing a single Scope for all VMs with the same lifetime and code (stats) or having a separate Scope for each.    We do create a separate Scope for VMs with different lifetimes and code already.

Reply:
When do you *want* to reap stale stats?

Reply:
I would prefer to reap them when the last VM using them exists.  *If* we have to add support for clearing the stats periodically on running VMs I am willing to do so, but I would rather pursue other ways to limit memory  growth first.  @mandarjog is requesting the reaping because of the memory usage, but I am hoping we can get it down without that feature.

Reply:
suggest dropping a TODO now to rethink the scoping policy. I think the current structure is not benefiting you.